File: api/src/store/issues.ts
import { readdirSync, readFileSync, existsSync, mkdirSync, writeFileSync, rmSync, statSync, unlinkSync } from "node:fs"
import path from "node:path"
import matter from "gray-matter"
import { parseFrontmatter } from "../lib/frontmatter.js"
import { paths } from "../lib/paths.js"

export type IssueStatus = "open" | "in-progress" | "closed"
export type IssueType = "bug" | "task" | "improvement" | "chore"
export type Priority = "critical" | "high" | "medium" | "low"

export interface Issue {
  slug: string
  title: string
  status: IssueStatus
  type: IssueType
  priority: Priority
  assignees: string[]
  points: number | null
  sprint: string | null
  feature: string | null
  labels: string[]
  created: string | null
  modified: string | null
  body: string
  order?: number
}

// Priority ordering for sorting (lower index = higher priority)
const PRIORITY_ORDER: Priority[] = ["critical", "high", "medium", "low"]
const STATUS_ORDER: IssueStatus[] = ["open", "in-progress", "closed"]
const TYPE_ORDER: IssueType[] = ["bug", "task", "improvement", "chore"]

function normalizePriority(raw: unknown): Priority {
  const s = String(raw ?? "").toLowerCase()
  if (PRIORITY_ORDER.includes(s as Priority)) return s as Priority
  const n = Number(raw)
  if (!isNaN(n)) {
    if (n <= 1) return "critical"
    if (n <= 3) return "high"
    if (n <= 6) return "medium"
    return "low"
  }
  return "medium"
}

function normalizeStatus(raw: unknown): IssueStatus {
  const s = String(raw ?? "").toLowerCase()
  if (STATUS_ORDER.includes(s as IssueStatus)) return s as IssueStatus
  return "open"
}

function normalizeType(raw: unknown): IssueType {
  const s = String(raw ?? "").toLowerCase()
  if (TYPE_ORDER.includes(s as IssueType)) return s as IssueType
  return "task"
}

function parseAssignees(raw: unknown): string[] {
  if (Array.isArray(raw)) return raw.map(String).filter(Boolean)
  if (raw && typeof raw === "string") return [raw]
  return []
}

function parseLabels(raw: unknown): string[] {
  if (Array.isArray(raw)) return raw.map(String).filter(Boolean)
  return []
}

function parseIssue(slug: string, data: Record<string, unknown>, content: string): Issue {
  return {
    slug,
    title: String(data.title ?? slug),
    status: normalizeStatus(data.status),
    type: normalizeType(data.type),
    priority: normalizePriority(data.priority),
    assignees: parseAssignees(data.assignee ?? data.assignees),
    points: data.points != null ? Number(data.points) : null,
    sprint: data.sprint ? String(data.sprint) : null,
    feature: data.feature ? String(data.feature) : null,
    labels: parseLabels(data.labels),
    created: data.created ? String(data.created) : null,
    modified: data.modified ? String(data.modified) : null,
    body: content,
    order: typeof data.order === "number" ? data.order : undefined,
  }
}

// ---------------------------------------------------------------------------
// Paginated listing
// ---------------------------------------------------------------------------

export interface ListIssuesParams {
  page?: number
  limit?: number
  status?: IssueStatus[]
  type?: IssueType[]
  priority?: Priority[]
  assignee?: string | null  // string = name, null = unassigned filter
  feature?: string | null   // string = feature slug, null = unlinked filter
  pointsMin?: number
  pointsMax?: number
  labels?: string[]
  sort?: "order" | "title" | "status" | "type" | "priority" | "modified"
  dir?: "asc" | "desc"
}

export interface PaginatedIssues {
  issues: Issue[]
  total: number
  page: number
  limit: number
  totalPages: number
}

function readAllIssues(): Issue[] {
  if (!existsSync(paths.issues)) return []

  return readdirSync(paths.issues, { withFileTypes: true })
    .filter((d) => d.isDirectory())
    .map((d) => {
      const slug = d.name
      const issuePath = path.join(paths.issues, slug, "ISSUE.md")
      if (!existsSync(issuePath)) return null

      const { data, content } = parseFrontmatter(issuePath)
      return parseIssue(slug, data, content)
    })
    .filter((i): i is Issue => i !== null)
}

export function listIssues(params: ListIssuesParams = {}): PaginatedIssues {
  const {
    page = 1,
    limit = 25,
    status,
    type,
    priority,
    assignee,
    feature,
    pointsMin,
    pointsMax,
    labels,
    sort = "order",
    dir = "asc",
  } = params

  let issues = readAllIssues()

  // ── Filters ──
  if (status && status.length > 0) {
    issues = issues.filter((i) => status.includes(i.status))
  }
  if (type && type.length > 0) {
    issues = issues.filter((i) => type.includes(i.type))
  }
  if (priority && priority.length > 0) {
    issues = issues.filter((i) => priority.includes(i.priority))
  }
  if (assignee !== undefined) {
    if (assignee === null) {
      issues = issues.filter((i) => i.assignees.length === 0)
    } else {
      issues = issues.filter((i) => i.assignees.includes(assignee))
    }
  }
  if (feature !== undefined) {
    if (feature === null) {
      issues = issues.filter((i) => i.feature === null)
    } else {
      issues = issues.filter((i) => i.feature === feature)
    }
  }
  if (pointsMin !== undefined) {
    issues = issues.filter((i) => i.points != null && i.points >= pointsMin)
  }
  if (pointsMax !== undefined) {
    issues = issues.filter((i) => i.points != null && i.points <= pointsMax)
  }
  if (labels && labels.length > 0) {
    issues = issues.filter((i) =>
      labels.some((l) => i.labels.includes(l))
    )
  }

  // ── Sort ──
  issues.sort((a, b) => {
    let cmp = 0
    switch (sort) {
      case "order": {
        const oa = a.order ?? Infinity
        const ob = b.order ?? Infinity
        if (oa !== ob) cmp = oa - ob
        else cmp = a.title.localeCompare(b.title)
        break
      }
      case "title":
        cmp = a.title.localeCompare(b.title)
        break
      case "status":
        cmp = STATUS_ORDER.indexOf(a.status) - STATUS_ORDER.indexOf(b.status)
        break
      case "type":
        cmp = TYPE_ORDER.indexOf(a.type) - TYPE_ORDER.indexOf(b.type)
        break
      case "priority":
        cmp = PRIORITY_ORDER.indexOf(a.priority) - PRIORITY_ORDER.indexOf(b.priority)
        break
      case "modified":
        cmp = (a.modified ?? "").localeCompare(b.modified ?? "")
        break
    }
    return dir === "asc" ? cmp : -cmp
  })

  const total = issues.length
  const totalPages = Math.max(1, Math.ceil(total / limit))
  const safePage = Math.min(Math.max(1, page), totalPages)
  const start = (safePage - 1) * limit
  const paged = issues.slice(start, start + limit)

  return { issues: paged, total, page: safePage, limit, totalPages }
}

export function getIssue(slug: string): Issue | null {
  const issuePath = path.join(paths.issues, slug, "ISSUE.md")
  if (!existsSync(issuePath)) return null

  const { data, content } = parseFrontmatter(issuePath)
  return parseIssue(slug, data, content)
}

// ---------------------------------------------------------------------------
// Create
// ---------------------------------------------------------------------------

function slugify(title: string): string {
  return title
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "")
}

export interface CreateIssueInput {
  title: string
  body?: string
  type?: IssueType
  priority?: Priority
  feature?: string
}

export function createIssue(input: CreateIssueInput): Issue {
  const slug = slugify(input.title)
  if (!slug) throw new Error("Invalid issue title")

  const issueDir = path.join(paths.issues, slug)
  if (existsSync(issueDir)) {
    throw new Error(`Issue '${slug}' already exists`)
  }

  // Ensure parent issues directory exists
  if (!existsSync(paths.issues)) {
    mkdirSync(paths.issues, { recursive: true })
  }

  mkdirSync(issueDir, { recursive: true })

  const today = new Date().toISOString().slice(0, 10)
  const frontmatter: Record<string, unknown> = {
    title: input.title,
    status: "open",
    type: input.type ?? "task",
    priority: input.priority ?? "medium",
    created: today,
    modified: today,
  }
  if (input.feature) {
    frontmatter.feature = input.feature
  }

  const body = input.body ?? ""
  const content = matter.stringify(body ? `\n${body}\n` : "", frontmatter)
  const issuePath = path.join(issueDir, "ISSUE.md")
  writeFileSync(issuePath, content, "utf-8")

  return {
    slug,
    title: input.title,
    status: "open",
    type: input.type ?? "task",
    priority: input.priority ?? "medium",
    assignees: [],
    points: null,
    sprint: null,
    feature: input.feature ?? null,
    labels: [],
    created: today,
    modified: today,
    body: body ? `\n${body}\n` : "",
  }
}

// ---------------------------------------------------------------------------
// Update
// ---------------------------------------------------------------------------

export interface UpdateIssueInput {
  title?: string
  status?: IssueStatus
  type?: IssueType
  priority?: Priority
  assignees?: string[]
  points?: number | null
  sprint?: string | null
  feature?: string | null
  labels?: string[]
  body?: string
}

export function updateIssue(slug: string, input: UpdateIssueInput): Issue | null {
  const existing = getIssue(slug)
  if (!existing) return null

  const today = new Date().toISOString().slice(0, 10)

  const frontmatter: Record<string, unknown> = {
    title: input.title ?? existing.title,
    status: input.status ?? existing.status,
    type: input.type ?? existing.type,
    priority: input.priority ?? existing.priority,
    modified: today,
  }

  // Preserve created date
  if (existing.created) frontmatter.created = existing.created

  // Preserve order
  if (existing.order !== undefined) frontmatter.order = existing.order

  // Handle assignees (array)
  const assignees = input.assignees !== undefined ? input.assignees : existing.assignees
  if (assignees.length > 0) frontmatter.assignees = assignees

  // Handle points (explicitly nullable)
  const points = input.points !== undefined ? input.points : existing.points
  if (points != null) frontmatter.points = points

  // Handle sprint
  const sprint = input.sprint !== undefined ? input.sprint : existing.sprint
  if (sprint) frontmatter.sprint = sprint

  // Handle feature (explicitly nullable)
  const feature = input.feature !== undefined ? input.feature : existing.feature
  if (feature) frontmatter.feature = feature

  // Handle labels
  const labels = input.labels !== undefined ? input.labels : existing.labels
  if (labels.length > 0) frontmatter.labels = labels

  const body = input.body !== undefined ? input.body : existing.body
  const content = matter.stringify(body ? `\n${body}\n` : "", frontmatter)
  const issuePath = path.join(paths.issues, slug, "ISSUE.md")
  writeFileSync(issuePath, content, "utf-8")

  return {
    slug,
    title: frontmatter.title as string,
    status: frontmatter.status as IssueStatus,
    type: frontmatter.type as IssueType,
    priority: frontmatter.priority as Priority,
    assignees,
    points: points ?? null,
    sprint: sprint ?? null,
    feature: feature ?? null,
    labels,
    created: existing.created,
    modified: today,
    body: body ? `\n${body}\n` : "",
    order: existing.order,
  }
}

// ---------------------------------------------------------------------------
// Reorder
// ---------------------------------------------------------------------------

export function reorderIssues(slugs: string[], offset: number = 0): void {
  for (let i = 0; i < slugs.length; i++) {
    const slug = slugs[i]
    const issue = getIssue(slug)
    if (!issue) continue

    const newOrder = offset + i

    const frontmatter: Record<string, unknown> = {
      title: issue.title,
      status: issue.status,
      type: issue.type,
      priority: issue.priority,
      modified: issue.modified,
      order: newOrder,
    }
    if (issue.created) frontmatter.created = issue.created
    if (issue.assignees.length > 0) frontmatter.assignees = issue.assignees
    if (issue.points != null) frontmatter.points = issue.points
    if (issue.sprint) frontmatter.sprint = issue.sprint
    if (issue.feature) frontmatter.feature = issue.feature
    if (issue.labels.length > 0) frontmatter.labels = issue.labels

    const content = matter.stringify(issue.body ? `\n${issue.body}\n` : "", frontmatter)
    const issuePath = path.join(paths.issues, slug, "ISSUE.md")
    writeFileSync(issuePath, content, "utf-8")
  }
}

// ---------------------------------------------------------------------------
// Delete
// ---------------------------------------------------------------------------

export function deleteIssue(slug: string): boolean {
  const issueDir = path.join(paths.issues, slug)
  if (!existsSync(issueDir)) return false
  rmSync(issueDir, { recursive: true, force: true })
  return true
}

// ---------------------------------------------------------------------------
// Artifacts (supporting files)
// ---------------------------------------------------------------------------

export interface IssueArtifact {
  name: string
  size: number
  type: string
}

function guessFileType(filename: string): string {
  const ext = path.extname(filename).toLowerCase()
  const types: Record<string, string> = {
    ".md": "markdown",
    ".txt": "text",
    ".png": "image",
    ".jpg": "image",
    ".jpeg": "image",
    ".gif": "image",
    ".svg": "image",
    ".webp": "image",
    ".pdf": "document",
    ".json": "data",
    ".yaml": "data",
    ".yml": "data",
    ".csv": "data",
  }
  return types[ext] ?? "file"
}

export function listIssueArtifacts(slug: string): IssueArtifact[] | null {
  const issueDir = path.join(paths.issues, slug)
  if (!existsSync(issueDir)) return null

  return readdirSync(issueDir, { withFileTypes: true })
    .filter((entry) => entry.isFile() && entry.name !== "ISSUE.md")
    .map((entry) => {
      const filePath = path.join(issueDir, entry.name)
      const stats = statSync(filePath)
      return {
        name: entry.name,
        size: stats.size,
        type: guessFileType(entry.name),
      }
    })
    .sort((a, b) => a.name.localeCompare(b.name))
}

// ---------------------------------------------------------------------------
// Artifact content operations
// ---------------------------------------------------------------------------

const TEXT_TYPES = new Set(["markdown", "text", "data"])

const MIME_TYPES: Record<string, string> = {
  ".md": "text/markdown",
  ".txt": "text/plain",
  ".json": "application/json",
  ".yaml": "application/x-yaml",
  ".yml": "application/x-yaml",
  ".csv": "text/csv",
  ".png": "image/png",
  ".jpg": "image/jpeg",
  ".jpeg": "image/jpeg",
  ".gif": "image/gif",
  ".svg": "image/svg+xml",
  ".webp": "image/webp",
  ".pdf": "application/pdf",
  ".html": "text/html",
  ".css": "text/css",
  ".js": "application/javascript",
  ".ts": "text/plain",
  ".tsx": "text/plain",
  ".jsx": "text/plain",
  ".xml": "application/xml",
  ".sh": "text/x-shellscript",
  ".py": "text/x-python",
  ".rb": "text/x-ruby",
  ".go": "text/x-go",
  ".rs": "text/x-rust",
  ".sql": "text/x-sql",
  ".graphql": "text/plain",
  ".env": "text/plain",
  ".toml": "text/plain",
  ".ini": "text/plain",
  ".conf": "text/plain",
  ".log": "text/plain",
}

function getMimeType(filename: string): string {
  const ext = path.extname(filename).toLowerCase()
  return MIME_TYPES[ext] ?? "application/octet-stream"
}

/**
 * Validate an artifact filename to prevent path traversal.
 * Returns the safe absolute path or null if invalid.
 */
function safeArtifactPath(slug: string, filename: string): string | null {
  if (!filename || filename.includes("..") || filename.includes("/") || filename.includes("\\") || filename.includes("\0")) {
    return null
  }
  if (filename === "ISSUE.md") return null

  const issueDir = path.join(paths.issues, slug)
  const filePath = path.join(issueDir, filename)

  const resolved = path.resolve(filePath)
  const resolvedDir = path.resolve(issueDir)
  if (!resolved.startsWith(resolvedDir + path.sep)) return null

  return filePath
}

export function isTextArtifact(filename: string): boolean {
  const type = guessFileType(filename)
  if (TEXT_TYPES.has(type)) return true
  const ext = path.extname(filename).toLowerCase()
  const codeExts = new Set([
    ".js", ".ts", ".tsx", ".jsx", ".html", ".css",
    ".sh", ".py", ".rb", ".go", ".rs", ".sql",
    ".graphql", ".env", ".toml", ".ini", ".conf", ".log",
    ".xml",
  ])
  return codeExts.has(ext)
}

export interface ArtifactContent {
  name: string
  content: string
  type: string
  mimeType: string
  size: number
}

export function getArtifactContent(slug: string, filename: string): ArtifactContent | null {
  const filePath = safeArtifactPath(slug, filename)
  if (!filePath || !existsSync(filePath)) return null

  const stats = statSync(filePath)
  const content = readFileSync(filePath, "utf-8")
  return {
    name: filename,
    content,
    type: guessFileType(filename),
    mimeType: getMimeType(filename),
    size: stats.size,
  }
}

export function getArtifactFilePath(slug: string, filename: string): { filePath: string; mimeType: string } | null {
  const filePath = safeArtifactPath(slug, filename)
  if (!filePath || !existsSync(filePath)) return null
  return { filePath, mimeType: getMimeType(filename) }
}

export function saveArtifactContent(slug: string, filename: string, content: string): IssueArtifact | null {
  const issueDir = path.join(paths.issues, slug)
  if (!existsSync(issueDir)) return null

  const filePath = safeArtifactPath(slug, filename)
  if (!filePath) return null

  writeFileSync(filePath, content, "utf-8")
  const stats = statSync(filePath)
  return {
    name: filename,
    size: stats.size,
    type: guessFileType(filename),
  }
}

export function saveArtifactBuffer(slug: string, filename: string, buffer: Buffer): IssueArtifact | null {
  const issueDir = path.join(paths.issues, slug)
  if (!existsSync(issueDir)) return null

  const filePath = safeArtifactPath(slug, filename)
  if (!filePath) return null

  writeFileSync(filePath, buffer)
  const stats = statSync(filePath)
  return {
    name: filename,
    size: stats.size,
    type: guessFileType(filename),
  }
}

export function deleteArtifact(slug: string, filename: string): boolean {
  const filePath = safeArtifactPath(slug, filename)
  if (!filePath || !existsSync(filePath)) return false
  unlinkSync(filePath)
  return true
}

File: commandline/internal/store/issues.go
package store

import (
	"math"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"github.com/briantol/folio/internal/frontmatter"
	"github.com/briantol/folio/internal/model"
)

// IssueStore provides filesystem-backed CRUD for issues.
type IssueStore struct {
	paths *Paths
}

// NewIssueStore creates a new IssueStore.
func NewIssueStore(p *Paths) *IssueStore {
	return &IssueStore{paths: p}
}

func (s *IssueStore) readIssue(slug string) (*model.Issue, error) {
	mdPath := filepath.Join(s.paths.Issues, slug, "ISSUE.md")
	data, err := os.ReadFile(mdPath)
	if err != nil {
		return nil, err
	}

	doc, err := frontmatter.Parse(data)
	if err != nil {
		return nil, err
	}

	issue := &model.Issue{
		Slug:     slug,
		Title:    frontmatter.GetString(doc.Data, "title"),
		Status:   frontmatter.GetString(doc.Data, "status"),
		Type:     frontmatter.GetString(doc.Data, "type"),
		Priority: normalizePriority(doc.Data),
		Points:   frontmatter.GetFloat(doc.Data, "points"),
		Sprint:   frontmatter.GetStringPtr(doc.Data, "sprint"),
		Feature:  frontmatter.GetStringPtr(doc.Data, "feature"),
		Created:  frontmatter.GetString(doc.Data, "created"),
		Modified: frontmatter.GetString(doc.Data, "modified"),
		Body:     doc.Body,
		Order:    frontmatter.GetInt(doc.Data, "order"),
	}

	assignees := frontmatter.GetStringSlice(doc.Data, "assignees")
	if len(assignees) == 0 {
		if a := frontmatter.GetString(doc.Data, "assignee"); a != "" {
			assignees = []string{a}
		}
	}
	if assignees == nil {
		assignees = []string{}
	}
	issue.Assignees = assignees

	labels := frontmatter.GetStringSlice(doc.Data, "labels")
	if labels == nil {
		labels = []string{}
	}
	issue.Labels = labels

	if issue.Status == "" {
		issue.Status = "open"
	}
	if issue.Type == "" {
		issue.Type = "task"
	}

	return issue, nil
}

func (s *IssueStore) readAllIssues() []model.Issue {
	entries, err := os.ReadDir(s.paths.Issues)
	if err != nil {
		return nil
	}

	issues := make([]model.Issue, 0, len(entries))
	for _, entry := range entries {
		if !entry.IsDir() || strings.HasPrefix(entry.Name(), ".") {
			continue
		}
		issue, err := s.readIssue(entry.Name())
		if err != nil {
			continue
		}
		issues = append(issues, *issue)
	}

	return issues
}

// List returns a paginated, filtered, sorted list of issues.
func (s *IssueStore) List(params model.ListIssuesParams) model.PaginatedIssues {
	all := s.readAllIssues()

	filtered := make([]model.Issue, 0, len(all))
	for _, issue := range all {
		if !matchIssueFilters(issue, params) {
			continue
		}
		filtered = append(filtered, issue)
	}

	sortIssues(filtered, params.Sort, params.Dir)

	total := len(filtered)
	totalPages := int(math.Max(1, math.Ceil(float64(total)/float64(params.Limit))))
	page := params.Page
	if page > totalPages {
		page = totalPages
	}
	start := (page - 1) * params.Limit
	end := start + params.Limit
	if end > total {
		end = total
	}
	if start > total {
		start = total
	}

	return model.PaginatedIssues{
		Issues:     filtered[start:end],
		Total:      total,
		Page:       page,
		Limit:      params.Limit,
		TotalPages: totalPages,
	}
}

func matchIssueFilters(issue model.Issue, p model.ListIssuesParams) bool {
	if len(p.Status) > 0 && !containsStr(p.Status, issue.Status) {
		return false
	}
	if len(p.Type) > 0 && !containsStr(p.Type, issue.Type) {
		return false
	}
	if len(p.Priority) > 0 && !containsStr(p.Priority, issue.Priority) {
		return false
	}
	if p.Assignee != nil {
		if *p.Assignee == "" {
			if len(issue.Assignees) > 0 {
				return false
			}
		} else {
			if !containsStr(issue.Assignees, *p.Assignee) {
				return false
			}
		}
	}
	if p.Feature != nil {
		if *p.Feature == "" {
			// __unlinked__ filter
			if issue.Feature != nil {
				return false
			}
		} else {
			if issue.Feature == nil || *issue.Feature != *p.Feature {
				return false
			}
		}
	}
	if p.PointsMin != nil && (issue.Points == nil || *issue.Points < *p.PointsMin) {
		return false
	}
	if p.PointsMax != nil && (issue.Points == nil || *issue.Points > *p.PointsMax) {
		return false
	}
	if len(p.Labels) > 0 {
		found := false
		for _, l := range p.Labels {
			if containsStr(issue.Labels, l) {
				found = true
				break
			}
		}
		if !found {
			return false
		}
	}
	return true
}

var issueStatusOrder = map[string]int{
	"in-progress": 0,
	"open":        1,
	"closed":      2,
}

var issueTypeOrder = map[string]int{
	"bug":         0,
	"task":        1,
	"improvement": 2,
	"chore":       3,
}

func sortIssues(issues []model.Issue, sortBy, dir string) {
	sort.SliceStable(issues, func(i, j int) bool {
		var less bool
		switch sortBy {
		case "title":
			less = strings.ToLower(issues[i].Title) < strings.ToLower(issues[j].Title)
		case "status":
			less = issueStatusOrder[issues[i].Status] < issueStatusOrder[issues[j].Status]
		case "type":
			less = issueTypeOrder[issues[i].Type] < issueTypeOrder[issues[j].Type]
		case "priority":
			less = priorityOrder[issues[i].Priority] < priorityOrder[issues[j].Priority]
		case "modified":
			less = issues[i].Modified < issues[j].Modified
		default:
			less = issues[i].Order < issues[j].Order
		}
		if dir == "desc" {
			return !less
		}
		return less
	})
}

// Get returns a single issue by slug.
func (s *IssueStore) Get(slug string) *model.Issue {
	issue, err := s.readIssue(slug)
	if err != nil {
		return nil
	}
	return issue
}

// Create creates a new issue.
func (s *IssueStore) Create(input model.CreateIssueInput) (*model.Issue, error) {
	slug := slugify(input.Title)
	dir := filepath.Join(s.paths.Issues, slug)
	if _, err := os.Stat(dir); err == nil {
		return nil, os.ErrExist
	}

	if err := os.MkdirAll(dir, 0755); err != nil {
		return nil, err
	}

	today := time.Now().Format("2006-01-02")
	data := map[string]interface{}{
		"title":    input.Title,
		"status":   "open",
		"type":     "task",
		"created":  today,
		"modified": today,
	}
	if input.Type != nil {
		data["type"] = *input.Type
	}
	if input.Priority != nil {
		data["priority"] = *input.Priority
	} else {
		data["priority"] = "medium"
	}
	if input.Feature != nil {
		data["feature"] = *input.Feature
	}

	body := ""
	if input.Body != nil {
		body = *input.Body
	}

	content, err := frontmatter.Stringify(data, body)
	if err != nil {
		os.RemoveAll(dir)
		return nil, err
	}

	mdPath := filepath.Join(dir, "ISSUE.md")
	if err := os.WriteFile(mdPath, content, 0644); err != nil {
		os.RemoveAll(dir)
		return nil, err
	}

	return s.readIssue(slug)
}

// Update updates an existing issue. Returns nil if not found.
func (s *IssueStore) Update(slug string, input model.UpdateIssueInput) (*model.Issue, error) {
	mdPath := filepath.Join(s.paths.Issues, slug, "ISSUE.md")
	rawData, err := os.ReadFile(mdPath)
	if err != nil {
		return nil, nil
	}

	doc, err := frontmatter.Parse(rawData)
	if err != nil {
		return nil, err
	}

	if input.Title != nil {
		doc.Data["title"] = *input.Title
	}
	if input.Status != nil {
		doc.Data["status"] = *input.Status
	}
	if input.Type != nil {
		doc.Data["type"] = *input.Type
	}
	if input.Priority != nil {
		doc.Data["priority"] = *input.Priority
	}
	if input.AssigneesSet {
		if input.Assignees == nil {
			doc.Data["assignees"] = []string{}
		} else {
			doc.Data["assignees"] = input.Assignees
		}
		delete(doc.Data, "assignee")
	}
	if input.PointsSet {
		if input.Points == nil {
			doc.Data["points"] = nil
		} else {
			doc.Data["points"] = *input.Points
		}
	}
	if input.SprintSet {
		if input.Sprint == nil {
			delete(doc.Data, "sprint")
		} else {
			doc.Data["sprint"] = *input.Sprint
		}
	}
	if input.FeatureSet {
		if input.Feature == nil {
			delete(doc.Data, "feature")
		} else {
			doc.Data["feature"] = *input.Feature
		}
	}
	if input.LabelsSet {
		if input.Labels == nil {
			doc.Data["labels"] = []string{}
		} else {
			doc.Data["labels"] = input.Labels
		}
	}

	body := doc.Body
	if input.Body != nil {
		body = *input.Body
	}

	doc.Data["modified"] = time.Now().Format("2006-01-02")

	content, err := frontmatter.Stringify(doc.Data, body)
	if err != nil {
		return nil, err
	}

	if err := os.WriteFile(mdPath, content, 0644); err != nil {
		return nil, err
	}

	return s.readIssue(slug)
}

// Delete removes an issue.
func (s *IssueStore) Delete(slug string) bool {
	dir := filepath.Join(s.paths.Issues, slug)
	if _, err := os.Stat(dir); err != nil {
		return false
	}
	os.RemoveAll(dir)
	return true
}

// Reorder sets order for the given slugs.
func (s *IssueStore) Reorder(slugs []string, offset int) {
	for i, slug := range slugs {
		mdPath := filepath.Join(s.paths.Issues, slug, "ISSUE.md")
		rawData, err := os.ReadFile(mdPath)
		if err != nil {
			continue
		}
		doc, err := frontmatter.Parse(rawData)
		if err != nil {
			continue
		}
		doc.Data["order"] = offset + i
		content, err := frontmatter.Stringify(doc.Data, doc.Body)
		if err != nil {
			continue
		}
		os.WriteFile(mdPath, content, 0644)
	}
}

// ListArtifacts returns artifacts for an issue.
func (s *IssueStore) ListArtifacts(slug string) ([]model.Artifact, bool) {
	dir := filepath.Join(s.paths.Issues, slug)
	if _, err := os.Stat(dir); err != nil {
		return nil, false
	}

	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil, false
	}

	artifacts := make([]model.Artifact, 0)
	for _, entry := range entries {
		if entry.IsDir() || entry.Name() == "ISSUE.md" {
			continue
		}
		info, err := entry.Info()
		if err != nil {
			continue
		}
		artifacts = append(artifacts, model.Artifact{
			Name: entry.Name(),
			Size: info.Size(),
			Type: artifactType(entry.Name()),
		})
	}

	return artifacts, true
}

// GetArtifactContent returns the text content of an issue artifact.
func (s *IssueStore) GetArtifactContent(slug, filename string) *model.ArtifactContent {
	path, ok := s.safeArtifactPath(slug, filename)
	if !ok {
		return nil
	}
	data, err := os.ReadFile(path)
	if err != nil {
		return nil
	}
	mt := mimeType(filename)
	return &model.ArtifactContent{
		Name:     filename,
		Content:  string(data),
		Type:     artifactType(filename),
		MimeType: mt,
		Size:     int64(len(data)),
	}
}

// GetArtifactFilePath returns absolute path and mime type for raw serving.
func (s *IssueStore) GetArtifactFilePath(slug, filename string) (string, string, bool) {
	path, ok := s.safeArtifactPath(slug, filename)
	if !ok {
		return "", "", false
	}
	if _, err := os.Stat(path); err != nil {
		return "", "", false
	}
	return path, mimeType(filename), true
}

// SaveArtifactContent saves text content to an issue artifact.
func (s *IssueStore) SaveArtifactContent(slug, filename, content string) *model.Artifact {
	path, ok := s.safeArtifactPath(slug, filename)
	if !ok {
		return nil
	}
	if err := os.WriteFile(path, []byte(content), 0644); err != nil {
		return nil
	}
	info, _ := os.Stat(path)
	return &model.Artifact{
		Name: filename,
		Size: info.Size(),
		Type: artifactType(filename),
	}
}

// SaveArtifactBuffer saves binary content to an issue artifact.
func (s *IssueStore) SaveArtifactBuffer(slug, filename string, data []byte) *model.Artifact {
	path, ok := s.safeArtifactPath(slug, filename)
	if !ok {
		return nil
	}
	if err := os.WriteFile(path, data, 0644); err != nil {
		return nil
	}
	return &model.Artifact{
		Name: filename,
		Size: int64(len(data)),
		Type: artifactType(filename),
	}
}

// DeleteArtifact deletes an issue artifact.
func (s *IssueStore) DeleteArtifact(slug, filename string) bool {
	path, ok := s.safeArtifactPath(slug, filename)
	if !ok {
		return false
	}
	if err := os.Remove(path); err != nil {
		return false
	}
	return true
}

func (s *IssueStore) safeArtifactPath(slug, filename string) (string, bool) {
	if filename == "ISSUE.md" || filename == "" {
		return "", false
	}
	if strings.Contains(filename, "..") || strings.ContainsAny(filename, "/\\\x00") {
		return "", false
	}
	dir := filepath.Join(s.paths.Issues, slug)
	if _, err := os.Stat(dir); err != nil {
		return "", false
	}
	path := filepath.Join(dir, filename)
	resolved, err := filepath.Abs(path)
	if err != nil {
		return "", false
	}
	absDir, _ := filepath.Abs(dir)
	if !strings.HasPrefix(resolved, absDir+string(filepath.Separator)) {
		return "", false
	}
	return path, true
}
